diff --git a/migration/cpr.c b/migration/cpr.c
index 42ad0b0d500e..7bc61e85c393 100644
--- a/migration/cpr.c
+++ b/migration/cpr.c
@@ -130,10 +130,12 @@ static const VMStateDescription vmstate_cpr_state = {
         VMSTATE_QLIST_V(fds, CprState, 1, vmstate_cpr_fd, CprFd, next),
         VMSTATE_END_OF_LIST()
     },
+#ifdef CONFIG_VFIO
     .subsections = (const VMStateDescription * const []) {
         &vmstate_cpr_vfio_devices,
         NULL
     }
+#endif
 };
 /*************************************************************************/
 
