diff --git a/block/meson.build b/block/meson.build
index 34b1b2a30630..6c0eb34f3834 100644
--- a/block/meson.build
+++ b/block/meson.build
@@ -50,17 +50,17 @@ endif
 if get_option('vdi').allowed()
   block_ss.add(files('vdi.c'))
 endif
-if get_option('vhdx').allowed()
+if get_option('vhdx').allowed() and not xen.found()
   block_ss.add(files(
     'vhdx-endian.c',
     'vhdx-log.c',
     'vhdx.c'
   ))
 endif
-if get_option('vmdk').allowed()
+if get_option('vmdk').allowed() and not xen.found()
   block_ss.add(files('vmdk.c'))
 endif
-if get_option('vpc').allowed()
+if get_option('vpc').allowed() and not xen.found()
   block_ss.add(files('vpc.c'))
 endif
 if get_option('cloop').allowed()
diff --git a/configs/devices/i386-softmmu/default.mak b/configs/devices/i386-softmmu/default.mak
index bc0479a7e0a3..9ff00082f9c1 100644
--- a/configs/devices/i386-softmmu/default.mak
+++ b/configs/devices/i386-softmmu/default.mak
@@ -26,8 +26,37 @@
 #CONFIG_SGX=n
 
 # Boards are selected by default, uncomment to keep out of the build.
-# CONFIG_ISAPC=n
-# CONFIG_I440FX=n
-# CONFIG_Q35=n
-# CONFIG_MICROVM=n
-# CONFIG_NITRO_ENCLAVE=n
+CONFIG_ISAPC=n
+CONFIG_I440FX=y
+CONFIG_Q35=n
+CONFIG_MICROVM=n
+CONFIG_NITRO_ENCLAVE=n
+
+# Wanted devices:
+CONFIG_AHCI=y
+CONFIG_AHCI_ICH9=y
+CONFIG_BOCHS_DISPLAY=y
+CONFIG_E1000_PCI=y
+CONFIG_EDID=y
+CONFIG_FDC=y
+CONFIG_GENERIC_LOADER=y
+CONFIG_ISA_DEBUG=y
+CONFIG_NVDIMM=y
+CONFIG_NVME_PCI=y
+CONFIG_PARALLEL=y
+CONFIG_RTL8139_PCI=y
+CONFIG_TPM_EMULATOR=y
+CONFIG_TPM_CRB=y
+CONFIG_USB=y
+CONFIG_USB_EHCI=y
+CONFIG_USB_EHCI_PCI=y
+CONFIG_USB_OHCI=y
+CONFIG_USB_OHCI_PCI=y
+CONFIG_USB_TABLET_WACOM=y
+CONFIG_USB_UHCI=y
+CONFIG_USB_HID=y
+CONFIG_VGA_CIRRUS=y
+CONFIG_VGA_PCI=y
+
+# Extra devices to get build working (investigate if these are needed)
+CONFIG_SPDM_SOCKET=y
diff --git a/hw/meson.build b/hw/meson.build
index 791ce21ab42a..993a407ee543 100644
--- a/hw/meson.build
+++ b/hw/meson.build
@@ -38,7 +38,11 @@ subdir('tpm')
 subdir('uefi')
 subdir('ufs')
 subdir('usb')
-subdir('vfio')
+
+if not xen.found()
+  subdir('vfio')
+endif
+
 subdir('vfio-user')
 subdir('virtio')
 subdir('vmapple')
diff --git a/include/hw/boards.h b/include/hw/boards.h
index f94713e6e29a..2984d0aba4dc 100644
--- a/include/hw/boards.h
+++ b/include/hw/boards.h
@@ -642,8 +642,8 @@ struct MachineState {
  * and ver_machine_deletion_version logic in docs/conf.py and
  * the text in docs/about/deprecated.rst
  */
-#define MACHINE_VER_DELETION_MAJOR 6
-#define MACHINE_VER_DEPRECATION_MAJOR 3
+#define MACHINE_VER_DELETION_MAJOR 10
+#define MACHINE_VER_DEPRECATION_MAJOR 10
 
 /*
  * Expands to a static string containing a deprecation
diff --git a/migration/cpr.c b/migration/cpr.c
index 42ad0b0d500e..bfa192b35586 100644
--- a/migration/cpr.c
+++ b/migration/cpr.c
@@ -129,10 +129,12 @@ static const VMStateDescription vmstate_cpr_state = {
     .fields = (VMStateField[]) {
         VMSTATE_QLIST_V(fds, CprState, 1, vmstate_cpr_fd, CprFd, next),
         VMSTATE_END_OF_LIST()
+#ifdef CONFIG_VFIO
     },
     .subsections = (const VMStateDescription * const []) {
         &vmstate_cpr_vfio_devices,
         NULL
+#endif
     }
 };
 /*************************************************************************/
