diff --git a/hw/i386/xen/xen_platform.c b/hw/i386/xen/xen_platform.c
index 99fc57c6a2cf..624151eeb810 100644
--- a/hw/i386/xen/xen_platform.c
+++ b/hw/i386/xen/xen_platform.c
@@ -71,6 +71,7 @@ struct PCIXenPlatformState {
     char log_buffer[4096];
     int log_buffer_off;
 
+    uint16_t class_id;
     uint16_t device_id;
 };
 
@@ -567,6 +568,7 @@ static void xen_platform_realize(PCIDevice *dev, Error **errp)
     pci_conf[PCI_INTERRUPT_PIN] = 1;
 
     pci_set_word(pci_conf + PCI_DEVICE_ID, d->device_id);
+    pci_set_word(pci_conf + PCI_CLASS_DEVICE, d->class_id);
 
     platform_ioport_bar_setup(d);
     pci_register_bar(dev, 0, PCI_BASE_ADDRESS_SPACE_IO, &d->bar);
@@ -588,7 +590,9 @@ static void platform_reset(DeviceState *dev)
 
 static const Property xen_platform_props[] = {
     DEFINE_PROP_UINT16("device-id", PCIXenPlatformState, device_id,
-                       PCI_DEVICE_ID_XEN_PLATFORM)
+                       PCI_DEVICE_ID_XEN_PLATFORM),
+    DEFINE_PROP_UINT16("class-id", PCIXenPlatformState, class_id,
+                       PCI_CLASS_OTHERS << 8 | 0x80)
 };
 
 static void xen_platform_class_init(ObjectClass *klass, const void *data)
@@ -598,7 +602,6 @@ static void xen_platform_class_init(ObjectClass *klass, const void *data)
 
     k->realize = xen_platform_realize;
     k->vendor_id = PCI_VENDOR_ID_XEN;
-    k->class_id = PCI_CLASS_OTHERS << 8 | 0x80;
     k->subsystem_vendor_id = PCI_VENDOR_ID_XEN;
     k->subsystem_id = PCI_DEVICE_ID_XEN_PLATFORM;
     k->revision = 1;
