diff --git a/hw/display/vga.c b/hw/display/vga.c
index e2d5df2..05290d5 100644
--- a/hw/display/vga.c
+++ b/hw/display/vga.c
@@ -2296,4 +2296,7 @@ void vga_init_vbe(VGACommonState *s, Object *obj, MemoryRegion *system_memory)
                                 VBE_DISPI_LFB_PHYSICAL_ADDRESS,
                                 &s->vram_vbe);
     s->vbe_mapped = 1;
+
+    if (vgt_vga_enabled)
+        s->vbe_extended = true;
 }
