diff --git a/hw/display/cirrus_vga.c b/hw/display/cirrus_vga.c
index 5198037..6794a44 100644
--- a/hw/display/cirrus_vga.c
+++ b/hw/display/cirrus_vga.c
@@ -3062,7 +3062,7 @@ static void cirrus_vga_class_init(ObjectClass *klass, void *data)
     PCIDeviceClass *k = PCI_DEVICE_CLASS(klass);
 
     k->realize = pci_cirrus_vga_realize;
-    k->romfile = VGABIOS_CIRRUS_FILENAME;
+    k->romfile = NULL; //VGABIOS_CIRRUS_FILENAME;
     k->vendor_id = PCI_VENDOR_ID_CIRRUS;
     k->device_id = CIRRUS_ID_CLGD5446;
     k->class_id = PCI_CLASS_DISPLAY_VGA;
diff --git a/hw/display/vga-pci.c b/hw/display/vga-pci.c
index 1dfa331..b6fdeaa 100644
--- a/hw/display/vga-pci.c
+++ b/hw/display/vga-pci.c
@@ -345,7 +345,7 @@ static void vga_class_init(ObjectClass *klass, void *data)
     PCIDeviceClass *k = PCI_DEVICE_CLASS(klass);
 
     k->realize = pci_std_vga_realize;
-    k->romfile = "vgabios-stdvga.bin";
+    k->romfile = NULL; //"vgabios-stdvga.bin";
     k->class_id = PCI_CLASS_DISPLAY_VGA;
     dc->props = vga_pci_properties;
     dc->hotpluggable = false;
diff --git a/hw/net/rtl8139.c b/hw/net/rtl8139.c
index cb51613..1a58e1a 100644
--- a/hw/net/rtl8139.c
+++ b/hw/net/rtl8139.c
@@ -3535,7 +3535,7 @@ static void rtl8139_class_init(ObjectClass *klass, void *data)
 
     k->realize = pci_rtl8139_realize;
     k->exit = pci_rtl8139_uninit;
-    k->romfile = "efi-rtl8139.rom";
+    k->romfile = NULL; //"efi-rtl8139.rom";
     k->vendor_id = PCI_VENDOR_ID_REALTEK;
     k->device_id = PCI_DEVICE_ID_REALTEK_8139;
     k->revision = RTL8139_PCI_REVID; /* >=0x20 is for 8139C+ */
