diff --git a/util/qemu-sockets.c b/util/qemu-sockets.c
index 7c588dd6f1..ad6ea6f99b 100644
--- a/util/qemu-sockets.c
+++ b/util/qemu-sockets.c
@@ -244,6 +244,7 @@ static int inet_listen_saddr(InetSocketAddress *saddr,
                                 sizeof(v6only));
             }
 #endif
+            socket_set_fast_reuse(slisten);
             if (bind(slisten, e->ai_addr, e->ai_addrlen) == 0) {
                 goto listen;
             }
