diff --git a/Makefile.objs b/Makefile.objs
index 06f74b8..98effff 100644
--- a/Makefile.objs
+++ b/Makefile.objs
@@ -3,6 +3,7 @@
 stub-obj-y = stubs/ crypto/
 util-obj-y = util/ qobject/ qapi/
 util-obj-y += qmp-introspect.o qapi-types.o qapi-visit.o qapi-event.o
+util-obj-y += logging.o
 
 #######################################################################
 # block-obj-y is code used by both qemu system emulation and qemu-img
diff --git a/include/qemu-common.h b/include/qemu-common.h
index 1430390..a951308 100644
--- a/include/qemu-common.h
+++ b/include/qemu-common.h
@@ -17,6 +17,7 @@
 #define TFR(expr) do { if ((expr) != -1) break; } while (errno == EINTR)
 
 #include "qemu/option.h"
+#include "logging.h"
 
 /* Copyright string for -version arguments, About dialogs, etc */
 #define QEMU_COPYRIGHT "Copyright (c) 2003-2016 " \
diff --git a/logging.c b/logging.c
new file mode 100644
index 0000000..18e3116
--- /dev/null
+++ b/logging.c
@@ -0,0 +1,56 @@
+#include <stdarg.h>
+#include <stdio.h>
+#include <syslog.h>
+#include "logging.h"
+
+#undef fprintf
+#undef printf
+#undef vfprintf
+
+/* the use of logging_set_prefix() is optional */
+void logging_set_prefix(const char *ident)
+{
+    closelog();
+    openlog(ident, LOG_NOWAIT | LOG_PID, LOG_DAEMON);
+}
+
+static inline void __syslog_vfprintf(const char *format, va_list ap)
+{
+    vsyslog(LOG_DAEMON | LOG_NOTICE, format, ap);
+}
+
+int qemu_log_vfprintf(FILE *stream, const char *format, va_list ap)
+{
+    if (stream == stdout || stream == stderr) {
+        __syslog_vfprintf(format, ap);
+        return 0;
+    } else {
+        return vfprintf(stream, format, ap);
+    }
+}
+
+int qemu_log_printf(const char *format, ...)
+{
+    va_list ap;
+
+    va_start(ap, format);
+    __syslog_vfprintf(format, ap);
+    va_end(ap);
+
+    return 0;
+}
+
+int qemu_log_fprintf(FILE *stream, const char *format, ...)
+{
+    va_list ap;
+    int ret = 0;
+
+    va_start(ap, format);
+    if (stream == stdout || stream == stderr)
+        __syslog_vfprintf(format, ap);
+    else
+        ret = vfprintf(stream, format, ap);
+    va_end(ap);
+
+    return ret;
+}
diff --git a/logging.h b/logging.h
new file mode 100644
index 0000000..b0b2025
--- /dev/null
+++ b/logging.h
@@ -0,0 +1,29 @@
+#ifndef LOGGING_H_
+# define LOGGING_H_
+
+# include <stdarg.h>
+# include <stdio.h>
+
+#ifdef printf
+# undef printf
+#endif
+#ifdef vfprintf
+# undef vfprintf
+#endif
+#ifdef fprintf
+# undef fprintf
+#endif
+
+# define printf(...) qemu_log_printf(__VA_ARGS__)
+# define vfprintf(...) qemu_log_vfprintf(__VA_ARGS__)
+# define fprintf(...) qemu_log_fprintf(__VA_ARGS__)
+
+void logging_set_prefix(const char *ident);
+int qemu_log_vfprintf(FILE *stream, const char *format, va_list ap);
+int qemu_log_printf(const char *format, ...)
+  __attribute__ ((format (printf, 1, 2)));
+int qemu_log_fprintf(FILE *stream, const char *format, ...)
+  __attribute__ ((format (printf, 2, 3)));
+
+
+#endif /* !LOGGING_H_ */
diff --git a/vl.c b/vl.c
index d77dd86..75b788f 100644
--- a/vl.c
+++ b/vl.c
@@ -123,6 +123,7 @@ int main(int argc, char **argv)
 #include "sysemu/replay.h"
 #include "qapi/qmp/qerror.h"
 #include "sysemu/iothread.h"
+#include "logging.h"
 
 #define MAX_VIRTIO_CONSOLES 1
 #define MAX_SCLP_CONSOLES 1
@@ -3864,6 +3865,8 @@ int main(int argc, char **argv, char **envp)
             case QEMU_OPTION_name:
                 opts = qemu_opts_parse_noisily(qemu_find_opts("name"),
                                                optarg, true);
+                /* get qemu_name */
+                logging_set_prefix(qemu_opt_get(opts, "guest"));
                 if (!opts) {
                     exit(1);
                 }
