diff --git a/include/hw/xen/xen.h b/include/hw/xen/xen.h
index 84b9819..efe75a0 100644
--- a/include/hw/xen/xen.h
+++ b/include/hw/xen/xen.h
@@ -49,4 +49,7 @@ void xen_register_framebuffer(struct MemoryRegion *mr);
 
 void xen_unassigned_access(hwaddr addr, bool is_write, void *val,
                            unsigned size);
+
+void xenstore_set_guest_clipboard(const uint8_t *buf, size_t len);
+
 #endif /* QEMU_HW_XEN_H */
diff --git a/ui/vnc.c b/ui/vnc.c
index 2c28a59..dffcf2c 100644
--- a/ui/vnc.c
+++ b/ui/vnc.c
@@ -29,6 +29,7 @@
 #include "vnc-jobs.h"
 #include "trace.h"
 #include "hw/qdev.h"
+#include "hw/xen/xen.h"
 #include "sysemu/sysemu.h"
 #include "qemu/error-report.h"
 #include "qemu/sockets.h"
@@ -1587,6 +1588,30 @@ uint32_t read_u32(uint8_t *data, size_t offset)
 
 static void client_cut_text(VncState *vs, size_t len, uint8_t *text)
 {
+    xenstore_set_guest_clipboard(text, len);
+}
+
+static void vnc_set_clipboard(VncState *vs, const uint8_t *buf, unsigned int len)
+{
+    char pad[3] = { 0, 0, 0 };
+    vnc_write_u8(vs, VNC_MSG_SERVER_CUT_TEXT); /* ServerCutText */
+    vnc_write(vs, pad, 3);   /* padding */
+    vnc_write_u32(vs, len);  /* length */
+    vnc_write(vs, buf, len); /* data */
+    vnc_flush(vs);
+}
+
+void vnc_dpy_set_clipboard(const uint8_t *buf, unsigned int len)
+{
+    VncDisplay *vd;
+
+    QTAILQ_FOREACH(vd, &vnc_displays, next) {
+        VncState *vs;
+
+        QTAILQ_FOREACH(vs, &vd->clients, next) {
+            vnc_set_clipboard(vs, buf, len);
+        }
+    }
 }
 
 static void check_pointer_type_change(Notifier *notifier, void *data)
diff --git a/ui/vnc.h b/ui/vnc.h
index d20b154..0722a31 100644
--- a/ui/vnc.h
+++ b/ui/vnc.h
@@ -574,4 +574,6 @@ int vnc_zrle_send_framebuffer_update(VncState *vs, int x, int y, int w, int h);
 int vnc_zywrle_send_framebuffer_update(VncState *vs, int x, int y, int w, int h);
 void vnc_zrle_clear(VncState *vs);
 
+void vnc_dpy_set_clipboard(const uint8_t *buf, unsigned int len);
+
 #endif /* QEMU_VNC_H */
diff --git a/xen-hvm.c b/xen-hvm.c
index fb39579..6a01c2f 100644
--- a/xen-hvm.c
+++ b/xen-hvm.c
@@ -31,6 +31,8 @@
 #include <xen/hvm/params.h>
 #include <xen/hvm/e820.h>
 
+#include "ui/vnc.h"
+
 //#define DEBUG_XEN_HVM
 
 #ifdef DEBUG_XEN_HVM
@@ -1209,6 +1211,185 @@ static void xen_wakeup_notifier(Notifier *notifier, void *data)
     xc_set_hvm_param(xen_xc, xen_domid, HVM_PARAM_ACPI_S_STATE, 0);
 }
 
+static struct ClipboardState {
+    struct xs_handle *xenstore;
+    uint8_t *tx_buf;  /* Data coming from the guest and going to the VNC client. */
+    unsigned int tx_len;      /* Length of data in tx_buf. */
+    unsigned int tx_buf_size; /* Size of tx_buf. */
+    int tx_failed;    /* True if receiving from the guest fails. */
+    int need_reset;   /* True if the set_clipboard key needs clearing. */
+    uint8_t *rx_buf;  /* Data going from the VNC client to the guest. */
+    unsigned int rx_len;       /* Length of data in rx_buf. */
+    unsigned int rx_offset;    /* Offset into rx_buf. */
+} clipboard_state;
+
+/* The token used to identify the set_clipboard watch. */
+#define SET_CLIPBOARD_TOKEN "set_clipboard"
+
+/* The token used to identify the report_clipboard watch. */
+#define REPORT_CLIPBOARD_TOKEN "report_clipboard"
+
+/* The maximum amount of data stored in xenstore at one time. */
+#define CLIPBOARD_CHUNK_SIZE 1024
+
+/* Limit the total amount we can accumulate from the guest to 1 MiB. */
+#define REPORT_CLIPBOARD_MAX_LEN (1024 * 1024)
+
+/*
+ * Send data from the VNC client to the guest via xenstore.
+ * Something happened on the state machine used for setting the
+ * guest clipboard.
+ */
+static void set_clipboard_event(void)
+{
+    uint8_t *buf;
+    unsigned int amount;
+    int rc;
+    char *cb_path, *dom_path;
+
+    dom_path = xs_get_domain_path(clipboard_state.xenstore, xen_domid);
+    if (!dom_path)
+        return;
+    rc = asprintf(&cb_path, "%s/data/set_clipboard", dom_path);
+    free(dom_path);
+    if (rc == -1)
+        return;
+
+    buf = xs_read(clipboard_state.xenstore, XBT_NULL, cb_path, NULL);
+    if (buf) {
+        /* Still waiting for the guest to ack the last operation. */
+        free(buf);
+        free(cb_path);
+        return;
+    }
+
+    if (clipboard_state.need_reset) {
+        xs_write(clipboard_state.xenstore, XBT_NULL, cb_path, "", 0);
+        free(cb_path);
+        clipboard_state.need_reset = 0;
+        return;
+    }
+
+    if (!clipboard_state.rx_buf) {
+        free(cb_path);
+        return;
+    }
+    amount = MIN(CLIPBOARD_CHUNK_SIZE,
+                 clipboard_state.rx_len - clipboard_state.rx_offset);
+    xs_write(clipboard_state.xenstore,
+             XBT_NULL,
+             cb_path,
+             clipboard_state.rx_buf + clipboard_state.rx_offset,
+             amount);
+    free(cb_path);
+    if (amount != 0) {
+        clipboard_state.rx_offset += amount;
+    } else {
+        free(clipboard_state.rx_buf);
+        clipboard_state.rx_buf = NULL;
+        clipboard_state.rx_len = 0;
+        clipboard_state.rx_offset = 0;
+    }
+}
+
+/*
+ * Receive clipboard data from the guest and forward it over VNC when the data
+ * is complete.
+ * Something happened on the state machine used by the guest to update
+ * the VNC clipboard.
+ */
+static void report_clipboard_event(void)
+{
+    unsigned int len, buf_size;
+    uint8_t *chunk, *buf;
+    char *cb_path, *dom_path;
+    int rc;
+
+    dom_path = xs_get_domain_path(clipboard_state.xenstore, xen_domid);
+    if (!dom_path)
+        return;
+    rc = asprintf(&cb_path, "%s/data/report_clipboard", dom_path);
+    free(dom_path);
+    if (rc == -1)
+        return;
+
+    chunk = xs_read(clipboard_state.xenstore, XBT_NULL, cb_path, &len);
+    if (!chunk) {
+        free(cb_path);
+        return;
+    }
+    if (len == 0) {
+        if (clipboard_state.tx_buf && !clipboard_state.tx_failed)
+            vnc_dpy_set_clipboard(clipboard_state.tx_buf,
+                                  clipboard_state.tx_len);
+        free(clipboard_state.tx_buf);
+        clipboard_state.tx_failed = 0;
+        clipboard_state.tx_len = 0;
+        clipboard_state.tx_buf = NULL;
+        clipboard_state.tx_buf_size = 0;
+    } else {
+        if (len + clipboard_state.tx_len > clipboard_state.tx_buf_size) {
+            buf_size = ((len + clipboard_state.tx_len) + 63) & ~63;
+            if (buf_size > REPORT_CLIPBOARD_MAX_LEN)
+                buf = NULL;
+            else
+                buf = realloc(clipboard_state.tx_buf, buf_size);
+        } else {
+            buf = clipboard_state.tx_buf;
+            buf_size = clipboard_state.tx_buf_size;
+        }
+        if (buf) {
+            clipboard_state.tx_buf_size = buf_size;
+            memcpy(buf + clipboard_state.tx_len, chunk, len);
+            clipboard_state.tx_len += len;
+            clipboard_state.tx_buf = buf;
+        } else {
+            clipboard_state.tx_failed = 1;
+        }
+    }
+    free(chunk);
+
+    xs_rm(clipboard_state.xenstore, XBT_NULL, cb_path);
+    free(cb_path);
+}
+
+void xenstore_set_guest_clipboard(const uint8_t *buf, size_t len)
+{
+    uint8_t *dup;
+
+    if (clipboard_state.rx_buf)
+        clipboard_state.need_reset = 1;
+    dup = malloc(len);
+    if (!dup)
+        return;
+    memcpy(dup, buf, len);
+
+    clipboard_state.rx_len = len;
+    clipboard_state.rx_offset = 0;
+    free(clipboard_state.rx_buf);
+    clipboard_state.rx_buf = dup;
+
+    set_clipboard_event();
+}
+
+static void xenstore_process_event(void *unused)
+{
+    char **vec;
+    unsigned int count;
+
+    vec = xs_read_watch(clipboard_state.xenstore, &count);
+    if (!vec)
+        goto out;
+
+    if (!strcmp(vec[XS_WATCH_TOKEN], SET_CLIPBOARD_TOKEN))
+        set_clipboard_event();
+    else if (!strcmp(vec[XS_WATCH_TOKEN], REPORT_CLIPBOARD_TOKEN))
+        report_clipboard_event();
+
+out:
+    free(vec);
+}
+
 void xen_hvm_init(PCMachineState *pcms, MemoryRegion **ram_memory)
 {
     int i, rc;
@@ -1216,6 +1397,7 @@ void xen_hvm_init(PCMachineState *pcms, MemoryRegion **ram_memory)
     xen_pfn_t bufioreq_pfn;
     evtchn_port_t bufioreq_evtchn;
     XenIOState *state;
+    char *dom_path, *watch_path;
 
     state = g_malloc0(sizeof (XenIOState));
 
@@ -1345,6 +1527,26 @@ void xen_hvm_init(PCMachineState *pcms, MemoryRegion **ram_memory)
     xen_be_register_common();
     xen_read_physmap(state);
 
+    dom_path = xs_get_domain_path(state->xenstore, xen_domid);
+
+    if (asprintf(&watch_path, "%s/data/set_clipboard", dom_path) == -1)
+        goto err;
+    if (!xs_watch(state->xenstore, watch_path, SET_CLIPBOARD_TOKEN))
+        goto err;
+    free(watch_path);
+
+    if (asprintf(&watch_path, "%s/data/report_clipboard", dom_path) == -1)
+        goto err;
+    if (!xs_watch(state->xenstore, watch_path, REPORT_CLIPBOARD_TOKEN))
+        goto err;
+    free(watch_path);
+
+    free(dom_path);
+
+    qemu_set_fd_handler(xs_fileno(state->xenstore),
+                        xenstore_process_event, NULL, NULL);
+    clipboard_state.xenstore = state->xenstore;
+
     /* Disable ACPI build because Xen handles it */
     pcms->acpi_build_enabled = false;
 
