From 0256d2995beb5c366922a10ca08ba9051fd651e3 Mon Sep 17 00:00:00 2001
From: Anoob Soman <anoob.soman@citrix.com>
Date: Tue, 15 Aug 2017 15:56:24 +0100
Subject: [PATCH] xen_pt_gfx: Don't load VGA rom from IGD device

Most often VGA rom doesn't exist or disabled.
diff --git a/hw/xen/xen_pt.c b/hw/xen/xen_pt.c
index 4750e2c..9d5ac87 100644
--- a/hw/xen/xen_pt.c
+++ b/hw/xen/xen_pt.c
@@ -915,15 +915,6 @@ static void xen_pt_realize(PCIDevice *d, Error **errp)
             return;
         }
 
-        xen_pt_setup_vga(s, &s->real_device, &err);
-        if (err) {
-            error_append_hint(&err, "Setup VGA BIOS of passthrough"
-                    " GFX failed");
-            error_propagate(errp, err);
-            xen_host_pci_device_put(&s->real_device);
-            return;
-        }
-
         /* Register ISA bridge for passthrough GFX. */
         xen_igd_passthrough_isa_bridge_create(s, &s->real_device);
     }
diff --git a/hw/xen/xen_pt.h b/hw/xen/xen_pt.h
index 1f043b4..d860ff5 100644
--- a/hw/xen/xen_pt.h
+++ b/hw/xen/xen_pt.h
@@ -342,6 +342,4 @@ static inline bool is_igd_vga_passthrough(XenHostPCIDevice *dev)
 }
 int xen_pt_register_vga_regions(XenPCIPassthroughState *s);
 int xen_pt_unregister_vga_regions(XenPCIPassthroughState *s);
-void xen_pt_setup_vga(XenPCIPassthroughState *s, XenHostPCIDevice *dev,
-                     Error **errp);
 #endif /* XEN_PT_H */
diff --git a/hw/xen/xen_pt_graphics.c b/hw/xen/xen_pt_graphics.c
index 91faa22..1c2defd 100644
--- a/hw/xen/xen_pt_graphics.c
+++ b/hw/xen/xen_pt_graphics.c
@@ -160,14 +160,6 @@ int xen_pt_unregister_vga_regions(XenPCIPassthroughState *s)
     return 0;
 }
 
-static void *get_vgabios(XenPCIPassthroughState *s, int *size,
-                       XenHostPCIDevice *dev)
-{
-    return pci_assign_dev_load_option_rom(&s->dev, OBJECT(&s->dev), size,
-                                          dev->domain, dev->bus,
-                                          dev->dev, dev->func);
-}
-
 /* Refer to Seabios. */
 struct rom_header {
     uint16_t signature;
@@ -194,52 +186,6 @@ struct pci_data {
     uint16_t reserved;
 } __attribute__((packed));
 
-void xen_pt_setup_vga(XenPCIPassthroughState *s, XenHostPCIDevice *dev,
-                     Error **errp)
-{
-    unsigned char *bios = NULL;
-    struct rom_header *rom;
-    int bios_size;
-    char *c = NULL;
-    char checksum = 0;
-    uint32_t len = 0;
-    struct pci_data *pd = NULL;
-
-    if (!is_igd_vga_passthrough(dev)) {
-        error_setg(errp, "Need to enable igd-passthrough");
-        return;
-    }
-
-    bios = get_vgabios(s, &bios_size, dev);
-    if (!bios) {
-        error_setg(errp, "VGA: Can't get VBIOS");
-        return;
-    }
-
-    /* Currently we fixed this address as a primary. */
-    rom = (struct rom_header *)bios;
-    pd = (void *)(bios + (unsigned char)rom->pcioffset);
-
-    /* We may need to fixup Device Identification. */
-    if (pd->device != s->real_device.device_id) {
-        pd->device = s->real_device.device_id;
-
-        len = rom->size * 512;
-        /* Then adjust the bios checksum */
-        for (c = (char *)bios; c < ((char *)bios + len); c++) {
-            checksum += *c;
-        }
-        if (checksum) {
-            bios[len - 1] -= checksum;
-            XEN_PT_LOG(&s->dev, "vga bios checksum is adjusted %x!\n",
-                       checksum);
-        }
-    }
-
-    /* Currently we fixed this address as a primary for legacy BIOS. */
-    cpu_physical_memory_rw(0xc0000, bios, bios_size, 1);
-}
-
 uint32_t igd_read_opregion(XenPCIPassthroughState *s)
 {
     uint32_t val = 0;
